### PR TITLE
Fix build

### PR DESCRIPTION
**Pull request by florian-zeidler**
@dfreudenberger We have now beaker tests and the module is able to install Solr in the default version.
@rebuy-de/it-platform

### CODE REVIEW

**Comment by bjoernhaeuser** [comment]
@das-vinculum please squash your commits :)

**Comment by das-vinculum** [comment]
@bjoernhaeuser Too slow :)

File: Vagrantfile [hunk 1]
```diff
@@ -0,0 +1,77 @@
+# -*- mode: ruby -*-
```
Comment (bjoernhaeuser) [inline Vagrantfile]: remove that one?

File: manifests/extract_file.pp [hunk 2]
```diff
@@ -0,0 +1,13 @@
+define solr5::extract_file( 
+$path_to_archive,
```
Comment (bjoernhaeuser) [inline manifests/extract_file.pp]: einrückung stimmt hier nicht :)

File: manifests/install.pp [hunk 3]
```diff
@@ -1,13 +1,24 @@
...
+
+  wget::fetch { "download ${package_url}":
+    source      => $package_url,
+    destination => "${package_target_dir}/${solr_archive_file_name}"
+  } ->
+
+  solr5::extract_file { 'install_solr_service.sh':
+     path_to_archive  => "${package_target_dir}",
+     archive_filename => "${solr_archive_file_name}",
+     file_to_extract  =>'install_solr_service.sh'
```
Comment (bjoernhaeuser) [inline manifests/install.pp]: einrückung :)

File: Vagrantfile (hunk 1 above)
Comment (florian-zeidler) [inline Vagrantfile]: No, it tells vim what's in this file.

Comment (bjoernhaeuser) [inline Vagrantfile]: No I meant remove the vagrant file ;)

**Comment by florian-zeidler** [comment]
Na, you can use that to quickly deploy / test the module.

File: Gemfile [hunk 4]
```diff
@@ -1,7 +1,18 @@
 source 'https://rubygems.org'
-
+ruby '1.9.3', :patchlevel => '551'
 puppetversion = ENV.key?('PUPPET_VERSION') ? "= #{ENV['PUPPET_VERSION']}" : ['>= 3.3']
-gem 'puppet', puppetversion
-gem 'puppetlabs_spec_helper', '>= 0.1.0'
-gem 'puppet-lint', '>= 0.3.2'
+gem 'metadata-json-lint'
+gem 'puppet', '3.8.3'
```
Comment (bjoernhaeuser) [inline Gemfile]: use puppetversion please

File: manifests/check_node_permissions.pp [hunk 5]
```diff
@@ -0,0 +1,5 @@
+define solr5::check_node_permissions($mode) {
+    exec { "/bin/chmod $mode $name":
```
Comment (bjoernhaeuser) [inline manifests/check_node_permissions.pp]: For what do we need it and why can't we use File {}?

File: spec/classes/solr5_spec.rb [hunk 6]
```diff
@@ -0,0 +1,14 @@
+require 'spec_helper'
+
+describe 'solr5', :type => 'class' do
+    
+    let :facts do
+        {
+            :concat_basedir => 'blubber',
```
Comment (bjoernhaeuser) [inline spec/classes/solr5_spec.rb]: :-)

File: spec/spec_helper.rb [hunk 7]
```diff
@@ -1 +1,8 @@
+require 'rubygems'
+require 'rspec-puppet'
 require 'puppetlabs_spec_helper/module_spec_helper'
+fixture_path = File.expand_path(File.join(__FILE__, '..', 'fixtures'))
+RSpec.configure do |c|
```
Comment (bjoernhaeuser) [inline spec/spec_helper.rb]: Hm, for what do you need this?

**Comment by dfreudenberger** [comment]
any reason we removed the dependency to download_archive and using wget and tar instead?

But I like having tests :+1:

**Comment by florian-zeidler** [comment]
I removed download_archive because it's intention is to download and extract the archive directly. The Solr installer cannot handle it this way, he needs the archive, which is at least currently stored under /tmp by download_archive, but is just unintentional.
So I thought it might be a good idea to use something, which is more intentional.

File: manifests/install.pp [hunk 8]
```diff
@@ -1,13 +1,24 @@
 class solr5::install inherits solr5 {
-  download_uncompress { $package_url:
-    distribution_name => $package_url,
-    dest_folder => $package_target_dir,
-    creates => $sources,
-    uncompress => 'tar.gz'
+
+  include wget
```
Comment (bjoernhaeuser) [inline manifests/install.pp]: why do you need to include wget?

Comment (florian-zeidler) [inline manifests/install.pp]: Because we need the archive completely unextracted. The solr installer is build that way. download_uncompress is actually leaving the archive in /tmp, but there is no guarantee for it that it stays that way.
Also I thought it would be irritating working on something which is not intentionally created.

**Comment by florian-zeidler** [comment]
@rebuy-de/it-platform I would like to get your remarks please.

File: Vagrantfile (hunk 1 above)
Comment (bjoernhaeuser) [inline Vagrantfile]: I dont see the point of having two ways to execute tests :-(

File: default.pp [hunk 9]
```diff
@@ -0,0 +1,4 @@
+
+class { 'java':} ->
```
Comment (bjoernhaeuser) [inline default.pp:2]: that file looks totally obsolete (you have to know that this is executed by the vagrantfile)

File: manifests/config.pp [hunk 10]
```diff
@@ -1,29 +1,38 @@
 class solr5::config inherits solr5 {
-  $config_file = "${solr_data_dir}/solr.in.sh"
+  $config_file = "${solr5::solr_data_dir}/solr.in.sh"
+
+  solr5::extract_file { 'solr.in.sh':
+    path_to_archive  => $solr5::package_target_dir,
+    archive_filename => $solr5::solr_archive_file_name,
+    file_to_extract  =>'solr.in.sh'
```
Comment (bjoernhaeuser) [inline manifests/config.pp:14]: Please add the comma in every line :)

File: manifests/extract_file.pp [hunk 11]
```diff
@@ -0,0 +1,12 @@
+define solr5::extract_file(
+  $path_to_archive,
+  $file_to_extract,
+  $archive_filename=$name,
+){
+  exec { $name:
+    command => "tar --extract --file ${path_to_archive}/${archive_filename} --wildcards --no-anchored '${file_to_extract}' -O > ${path_to_archive}/${file_to_extract} ",
+    cwd     => $path_to_archive,
+    path    => '/bin/',
```
Comment (bjoernhaeuser) [inline manifests/extract_file.pp:9]: Why not put the path into the command directly?

File: manifests/install.pp (hunk 8 above)
Comment (bjoernhaeuser) [inline manifests/install.pp]: Thats not what I am asking :) you already told me why you use the wget module. The question was "Why do you need that line here? What's the purpose of `include wget`. Do you need that line?". IMHO the module should work without explicitly including wget in the manifest.

File: manifests/params.pp [hunk 12]
```diff
@@ -1,7 +1,7 @@
 class solr5::params {
     $package_url = 'http://ftp.halifax.rwth-aachen.de/apache/lucene/solr/5.2.1/solr-5.2.1.tgz'
     $package_version = '5.2.1'
-    $package_target_dir = '/usr/local/src'
+    $package_target_dir = '/tmp'
```
Comment (bjoernhaeuser) [inline manifests/params.pp:4]: why this change?

File: spec/acceptance/nodesets/default.yml [hunk 13]
```diff
@@ -0,0 +1,14 @@
+HOSTS:
+  centos-6.6-64-nocm:
+    roles:
+      - agent
+    platform: el-6-x86_64
+    hypervisor: vagrant
+    box: puppetlabs/centos-6.6-64-nocm
+    box_url: https://vagrantcloud.com/puppetlabs/boxes/centos-6.6-64-nocm
```
Comment (bjoernhaeuser) [inline spec/acceptance/nodesets/default.yml]: you can remove the box_url, that will be auto-discovered

File: spec/acceptance/solr5_install_spec.rb [hunk 14]
```diff
@@ -0,0 +1,16 @@
+require 'spec_helper_acceptance'
+
+describe 'solr5 class' do
+
+    describe 'running puppet code' do
+
+        it 'should work with no errors' do
```
Comment (bjoernhaeuser) [inline spec/acceptance/solr5_install_spec.rb:7]: it `should install solr as a service and start it` 

and then add the necessary expectations :)

File: spec/classes/solr5_spec.rb [hunk 15]
```diff
@@ -0,0 +1,15 @@
+require 'spec_helper'
+
+describe 'solr5', :type => 'class' do
+
+  # You need this, because otherwise puppetlabs-concat fails
+  let :facts do
+    {
+       :concat_basedir => 'blubber',
```
Comment (bjoernhaeuser) [inline spec/classes/solr5_spec.rb]: That should be a better name.

File: spec/classes/solr5_spec.rb [hunk 16]
```diff
@@ -0,0 +1,15 @@
...
+     }
+  end
+
+  it{
+      should contain_class('solr5')
```
Comment (bjoernhaeuser) [inline spec/classes/solr5_spec.rb:14]: We should add more tests, right?

Comment (florian-zeidler) [inline spec/classes/solr5_spec.rb:14]: @bjoernhaeuser In general, you are right, but the first essential part is that the test starts a compilation of the whole thing.

File: manifests/install.pp (hunk 8 above)
Comment (florian-zeidler) [inline manifests/install.pp]: It turns out, you don't need that :exclamation:

File: manifests/params.pp (hunk 12 above)
Comment (florian-zeidler) [inline manifests/params.pp:4]: To unpack everything under /tmp, the software is installed later under /opt. I do not see a reason to keep the archive explicitly.

File: manifests/extract_file.pp (hunk 11 above)
Comment (florian-zeidler) [inline manifests/extract_file.pp:9]: Because the line is long enough even without the path, besides I think it adds just noise.

Comment (bjoernhaeuser) [inline manifests/extract_file.pp:9]: :+1:

**Comment by florian-zeidler** [comment]
@bjoernhaeuser @omares @dfreudenberger @rebuy-de/it-platform I would like to get this request closed today, so if you have any recommendations, feel free.

**Comment by dfreudenberger** [comment]
:+1:

File: .gitignore [hunk 17]
```diff
@@ -0,0 +1,3 @@
+Gemfile.lock
```
Comment (omares) [inline .gitignore]: also here i dont see a valid point to ignore the file

File: Rakefile [hunk 18]
```diff
@@ -1,18 +1,5 @@
...
-    sh "puppet parser validate --noop #{manifest}"
-  end
-  Dir['spec/**/*.rb','lib/**/*.rb'].each do |ruby_file|
-    sh "ruby -c #{ruby_file}" unless ruby_file =~ /spec\/fixtures/
-  end
-  Dir['templates/**/*.erb'].each do |template|
-    sh "erb -P -x -T '-' #{template} | ruby -c"
-  end
-end
+task :default => [:syntax,:spec_prep, :spec, :lint, :beaker]
```
Comment (omares) [inline Rakefile]: space after the first ,

File: manifests/config.pp [hunk 19]
```diff
@@ -1,16 +1,31 @@
-class solr5::config inherits solr5 {
+class solr5::config (
+  $solr_data_dir,
+  $package_target_dir,
+  $solr_archive_file_name,
+  $init_config,
+  $solr_user,
+  $solr_port
+)
```
Comment (omares) [inline manifests/config.pp]: Not consistent to other class definitions in the PR. One has the curly brace on the same line, the other has a space in between. Please pick one style

File: .gitignore (hunk 17 above)
Comment (florian-zeidler) [inline .gitignore]: It is just duplicated information, what should somebody do with it? In my opinion, the repository should only contain information, which can not be retrieved otherwise and the dependencies of the modules are fetched by bundler based on the Gemfile, we do not need this information twice. It is autogenerated stuff, I do insist not to commit something like that, you do not check in your IDE project files either.

Comment (omares) [inline .gitignore]: No it is not duplicate information. The file contains the exact version that you had installed at that certain point of time. It ensures package consistency. If someone else goes in and installs the gems he gets the ones with that exact same versions noted in the lock file instead of the most actual one matching your Gemfile definition. Removing the file could lead to unwanted behaviour because of inconsistent/unexpected environment.

Comment (omares) [inline .gitignore]: Also bundler itselfs recommends to check in the file http://bundler.io/v1.7/rationale.html#checking-your-code-into-version-control

Comment (florian-zeidler) [inline .gitignore]: Ok, well.

**Comment by bjoernhaeuser** [comment]
If you remove "blubber" and the Vagrantfile: :+1: :)
